RFC: Separate authorization and token endpoints

tux-rampage opened the pull request:
According to the [OAuth2 RFC 6749](https://tools.ietf.org/html/rfc6749)  the authorization and token endpoint  provide two different concerns and should be separated.

## Details

As of [section 3.1](https://tools.ietf.org/html/rfc6749#section-3.1), the authorization endpoint is designed
to Initiate an *interactive* authorization flow, where the user (resource owner) may be asked to authenticate against the authorization server as well as for consent to provide access to the requesting party. This endpoint will either redirect to the requesting party with an access code, the access token, or an error or should display an error page to the resource owner when the requesting party is invalid as defined in [section 3.1.2.4](https://tools.ietf.org/html/rfc6749#section-3.1.2.4). This endpoint *must* allow GET requests and *may* allow POST.

In contrast [section 3.2](https://tools.ietf.org/html/rfc6749#section-3.2), the token endpoint is designed for `automated` access token provision without user interaction. Usually this endpoint is used on the server side of the application to prevent access token exposure to the user agent.

With the current single middleware implementation consumers will face a bunch of issues:
* If the consumer wants to accept POST for the authorization endpoint (i.e. to post the consent decision of the resource owner), this middleware cannot be used as is.
* The consumer has no chance to intercept between validateAuthorizationRequest() and finishAuthorizationRequest() to provide the desired `UserEntityInterface` implementation or to set the approval flag.

Therefore this PR suggests to separate this MiddleWare (Which actually is a Request Handler) into 3 parts (Consider these classes are in namespace `Zend\Expressive\Authentication\OAuth2`):

* `TokenHandler` - Provides the implementation of the OAuth2 Token endpoint.
* `AuthorizatonMiddleware` - Prepares and validates the OAuth2 authrization request and passes it with the request object to the next handler(s)
* `AuthorizationHandler` Uses the `League\OAuth2\Server\RequestTypes\AuthorizationRequest` from the request and completes the response (proper redirecting to the requesting party)

It is up to the consumer to provide the User entity and approval information to the authorization request instance via a middleware. This gives consumers the flexibility to handle this step in any way they need to fulfill their business needs.

## Example:

```php
<?php

use Zend\Expressive\Authentication\OAuth2;

$app->route('/oauth2/authorize', [
    OAuth2\AuthorizatonMiddleware,
    Application\MyOAuthAuthorizationMiddleware::class, // <- this ensures authentication, populates the user and handles conscent
    OAuth2\AuthorizationHandler
], ['GET', 'POST']);
$app->route('/oauth2/token', OAuth2\TokenHandler::class, ['POST']);
```

- [x] Are you creating a new feature?
  - [x] Why is the new feature needed? What purpose does it serve?
  - [x] How will users use the new feature?
  - [x] Base your feature on the `develop` (:warning: missing) branch, and submit against that branch.
  - [x] Add only one feature per pull request; split multiple features over multiple pull requests
  - [x] Add tests for the new feature.
  - [x] Add documentation for the new feature.
  - [x] Add a `CHANGELOG.md` entry for the new feature.

- [x] Is this related to quality assurance?

tux-rampage committed:
Separate authorization and token endpoints

tux-rampage commented on src/AuthFlowFactoryTrait.php:
```diff
@@ -12,13 +12,15 @@
+ * Provides helper methods for authorization flow related factories
+ * @internal
+ */
+trait AuthFlowFactoryTrait
```
Even though this is discouraged by some, traits offer a solution to re-use code without building an inheritance hell. Plus, there is no official decision that forbids them.

This code is used in 3 factories and I don't see any benefit in duplicating this code three times which needs maintanence in three files if something needs to be changed in this code segment.

I don't know why git magically thinks this was moved from the `OAuth2MittlewareFactory`, it was not.

ezimuel replied:
I don't think we need this Trait. A simple `$container->get(AuthorizationServer::class)` is enough in each factory class. All the checks present in `AuthFlowFactoryTrait` are performed by the service manager by default.

tux-rampage replied:
Agree. Getting rid of the additional checks will simplify the codebase and increase maintainability.

I just ported it from the existing `OAuth2MiddlewareFactory`, because I didn't know if there was a specific intend for the specialized exception.

tux-rampage commented on src/AuthorizationHandler.php:
```diff
@@ -0,0 +1,54 @@
+ *
+ * @see https://tools.ietf.org/html/rfc6749#section-3.1.1
+ */
+class AuthorizationHandler implements RequestHandlerInterface
```
I decided to use this terminology and interface implementation, since it actually is a request handler.

ezimuel replied:
Ok

tux-rampage commented on src/AuthorizationHandler.php:
```diff
@@ -0,0 +1,54 @@
+
+    public function handle(ServerRequestInterface $request): ResponseInterface
+    {
+        $authRequest = $request->getAttribute(AuthorizationRequest::class);
```
This was created in the `AuthorizationMiddleware` and completed in the consumer's middleware.

tux-rampage commented on src/AuthorizationMiddleware.php:
```diff
@@ -0,0 +1,82 @@
+
+            // The next handler must take care of providing the
+            // authenticated user and the approval
+            $authRequest->setAuthorizationApproved(false);
```
Not sure if this is necessary, but it'll ensure unwanted approvals.

ezimuel replied:
I think this is necessary since here we are potentially delegating the authorization phase.

tux-rampage replied:
Indeed, better being explicit than relying on the 3rd party library's defaults.

tux-rampage commented on src/AuthorizationMiddleware.php:
```diff
@@ -0,0 +1,82 @@
+            // authenticated user and the approval
+            $authRequest->setAuthorizationApproved(false);
+
+            return $handler->handle($request->withAttribute(AuthorizationRequest::class, $authRequest));
```
Pass the the auth request via the server request object to the next handler to complete it.
(ensure authentication, giving consent, etc ...). This gives consumers the freedom to act according to their business needs

tux-rampage commented on src/AuthorizationMiddleware.php:
```diff
@@ -0,0 +1,82 @@
+            // The validation throws this exception if the request is not valid
+            // for example when the client id is invalid
+            return $exception->generateHttpResponse($response);
+        } catch (\Exception $exception) {
```
I am not sure if we should keep this. Maybe it's a better to let this fall through and let the error handler middleware picking it up.
Since this endpoint involves user interaction, the consumer can choose how to handle this situation.

ezimuel replied:
Yes, this is also something suggested by [thephpleague/oauth2-server](http://oauth2.thephpleague.com/authorization-server/auth-code-grant/#implementation).

tux-rampage commented on src/AuthorizationMiddlewareFactory.php:
```diff
@@ -0,0 +1,29 @@
+use Psr\Http\Message\ResponseInterface;
+use function sprintf;
+
+class AuthorizationMiddlewareFactory
```
Changed the name fom `OAuth2MiddlewareFactory` according to the middleware and deduplicated code.

tux-rampage commented on src/TokenEndpointHandler.php:
```diff
@@ -0,0 +1,83 @@
+
+        if (strtoupper($request->getMethod()) !== 'POST') {
+            return $response->withStatus(501); // Method not implemented
+        }
```
As for the Authorization endpoint, do we need this here, or should this be delegated to the
`route()` call? Anyways it should be consistent in both cases.

ezimuel replied:
I think we can remove this, since you have splitted the authorization endpoint from the token endpoint.

tux-rampage commented on src/OAuth2Middleware.php:
```diff
@@ -1,125 +0,0 @@
-            case 'POST':
-                return $this->accessTokenRequest($request);
-        }
-        return ($this->responseFactory)()->withStatus(501); // Method not implemented
```
I removed the method check from the `AuthorizationMiddleware` for the following reasons:

* POST may be allowed
* Imho it should be up to the consumer to restrict this, since even PUT might be thinkable depending on the consumer's needs.

ezimuel replied:
I agree.

tux-rampage reviewed (commented):
I annotated my intentions in the changeset

ezimuel commented:
@tux-rampage thanks for this PR, it's a great improvement from an architectural point of view.
I just reviewed it.
We need to fix and complete the uni tests. Let me know if you can complete the PR, otherwise I can work on it. Thanks again for your great work!

ezimuel reviewed (changes requested):
Please review my feedbacks, thanks!

tux-rampage commented:
@ezimuel Thanks for the feedback. I'll review and update the PR accordingly the next days.

wshafer commented:
Thanks for tackling this issue.  This was something I was also attempting to look at myself.   I'm glad you beat me to it.  ;)

One thing that I was having an issue with was the serialization and storage of the outh2 auth request object.  Perhaps I didn't see it, but how are you saving the request to be approved on approval requests, or how were you seeing this being addressed going forward?

tux-rampage committed:
Update test cases

tux-rampage committed:
Remove trait

tux-rampage committed:
Remove HTTP method check

tux-rampage committed:
Add missing service config

tux-rampage committed:
Fix coding style

tux-rampage commented:
@wshafer, I had an interesting chat with @xtreamwayz  in slack  about the issue.

I'd not recommend to serialize the OAuth2 Request Object. That would require you to have knowledge of iternals of Leaque's OAuth2 library. To me that looks like a not so good idea.

One option would be to keep the HTTP-Request parameters and Redirect to the authorization endpoint with these parameters when ready. In the OpenID provider that I am currently implementing, I use Angular as Frontend, so it's a one-pager and I can just render the UI view (parameterized) when user interaction is required. Additionally I have a middleware converting HTTP 301/302 redirects to JSON 200 responses and the redirect info in the response body so Angular will handle it.

tux-rampage commented:
I'll add some examples to the docs and drop a comment when this PR is ready for another review (there are still some tests missing).

tux-rampage committed:
Add auth handler test (wip)

tux-rampage committed:
Add missing authorization handler tests

tux-rampage committed:
Add token endpoint handler tests

tux-rampage committed:
fix coding style

tux-rampage committed:
Update docs

tux-rampage committed:
Remove unused imports

tux-rampage committed:
Add changelog entries

tux-rampage commented:
@ezimuel I've completed the PR, added docs and the CHANGELOG entries (feel free to rephrase - I'll apreciate any improvements). 

Sorry that this took a bit longer and thanks for your patience.

The drop of coverage is caused by the `@coversNothing` that I added to the [integration test](https://github.com/zendframework/zend-expressive-authentication-oauth2/pull/42/files#diff-39672decd272d5d1ffc439b2bc6ce1a0R57) (see http://phpunit.readthedocs.io/en/7.1/annotations.html#coversnothing and http://phpunit.readthedocs.io/en/7.1/code-coverage-analysis.html#specifying-covered-methods why). Before this the classes in Entities were reported to be test covered, which is actually not the case.

ezimuel commented:
@tux-rampage thanks for this great work! I'll review it asap.

tux-rampage commented:
@ezimuel Had an ytime to review this, yet? Let me know if I can support you.

ezimuel commented:
@tux-rampage thanks for the hard work!
